experiments/src/http_client.rs

```diff
@@ -7,16 +7,79 @@
 //! Working on the real Nimbus schema.
 
 use super::Experiment;
+use crate::Config;
 use anyhow::Result;
+use serde::de::SeqAccess;
+use serde_derive::Deserialize;
+use std::marker::PhantomData;
 use url::Url;
 use viaduct::{status_codes, Request, Response};
 
+const DEFAULT_BASE_URL: &str = "https://settings.stage.mozaws.net"; // TODO: Replace this with prod
+const DEFAULT_BUCKET_NAME: &str = "main";
+const DEFAULT_COLLECTION_NAME: &str = "messaging-experiments";
+
 // Making this a trait so that we can mock those later.
 pub(crate) trait SettingsClient {
     fn get_experiements_metadata(&self) -> Result<String>;
     fn get_experiments(&self) -> Result<Vec<Experiment>>;
 }
 
+#[derive(Deserialize)]
+struct ExperimentResponse {
+    #[serde(deserialize_with = "skip_invalid")]
+    data: Vec<Experiment>,
+}
+/// A custom deserializer that skips any invalid or null elements
+///
+/// This is needed to filter out experiments that do not abide by the
+/// schema. For production, we would probably need to remove this
+/// and ensure that the server contains *only* valid experiments
+fn skip_invalid<'de, D, T>(deserializer: D) -> Result<Vec<T>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+    T: serde::Deserialize<'de>,
+{
+    struct SkipInvalid<T>(PhantomData<T>);
+
+    impl<'de, T> serde::de::Visitor<'de> for SkipInvalid<T>
+    where
+        T: serde::Deserialize<'de>,
+    {
+        type Value = Vec<T>;
+
+        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+            formatter.write_str("Array with invalid elements")
+        }
+
+        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+        where
+            A: SeqAccess<'de>,
+        {
+            let mut vec = Vec::new();
+            // NOTE: In here we ignore any errors that appear in
+            // the deserialization process and instead
+            // skip any invalid elements
+            // In the future (when we have a source of *only* valid experiments)
+            // this should be removed
+            loop {
+                match seq.next_element::<T>() {
+                    Ok(inner) => {
+                        if let Some(inner) = inner {
+                            vec.push(inner);
+                        } else {
+                            break;
+                        }
+                    }
+                    Err(e) => log::warn!("Malformed experiment found! Error: {}", e),
```

Comment:
could we please include the name of the experiment here as well?

Reply:
Hmmm I'll take a look, may not be easy since all we receive back from `next_element` is the error. Should be possible with more playing around thou

Reply:
yea, feel free to time-box this as well. It's not critical, but makes it harder to find the broken experiment. We can also log it in a different statement afterwards, not sure of that helps.

Reply:
Figured out a way around this 😄